Gear collector

chloe6248 opened the pull request:
https://trello.com/c/COSCx6eP
https://trello.com/c/faL8QQnx

Commit:
Created gear collector subsystem, commands, and buttons

-Has two positions (in and out)
-Currently operates very similarly to the Shifter

Commit:
Comments added for GearCollector subsystem and related classes

Commit:
Added TODOs to OI and RobotMap

Commit:
Merge branch 'master' into GearCollector

# Conflicts:
#	src/org/usfirst/frc/team2399/robot/OI.java
#	src/org/usfirst/frc/team2399/robot/RobotMap.java

jameshochadel commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -86,13 +87,21 @@ public static double getRightY() {
 	// button.whenReleased(new ExampleCommand());
```
Remove commented-out code unless we have a really compelling reason to keep it

Commit:
Merge branch 'master' into GearCollector

# Conflicts:
#	src/org/usfirst/frc/team2399/robot/RobotMap.java

jameshochadel commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -86,13 +87,21 @@ public static double getRightY() {
-	 * OI Constructor
+	 * Presets - sets instances of commands and speeds for use with buttons
+	 */
+	private static Button gearOutButt = new JoystickButton(rightJoy, 55);
```
1. Fill in real numbers, and
2. Make them into constants

jameshochadel commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -86,13 +87,21 @@ public static double getRightY() {
+	/**OI Constructor
 	 * climbUpButt: While held down, go up at the preset speed (see above)
 	 * climbUpButt: While held down, go down at the preset speed (see above)
 	 * climbUpButt: When pressed, stop (emergency stop button)
```
All 3 above lines have "climbUpButt" before the colon. Is this correct?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -86,13 +87,21 @@ public static double getRightY() {
 	 */
-	public OI()
-	{
+	public OI(){
```
Nitpick: Add a space between parens and brackets, e.g. `public OI() {`

jameshochadel commented on src/org/usfirst/frc/team2399/robot/Robot.java:
```diff
@@ -32,6 +33,7 @@
+	public static GearCollector gearCollector = new GearCollector();
 
 	Command autonomousCommand;
 	SendableChooser<Command> chooser = new SendableChooser<>();
```
Why is second half of this not parameterized?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -33,22 +33,38 @@
 	public static final int CLIMBER_TALON_ADDRESS = 2399;
```
Need real address

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -26,29 +26,52 @@
 
+	/**
+	 * Ports for sensors, joysticks, and solenoids
+	 * TODO: Set actual ports
```
Is this TODO resolved?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -26,29 +26,52 @@
 	public static final int JOYDRIVE_LEFT_STICK_PORT = 0;
 	public static final int JOYDRIVE_RIGHT_STICK_PORT = 1;
 
+	public static final int GEAR_SOLENOID_PORT = 899;
```
Is this correct?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -26,29 +26,52 @@
-	 * TODO: Set actual forward constants for encoders
+	 * Forward Constants
+	 * To ensure positive/forward values of motion are what you expect.
+	 * TODO: Set actual values
```
Is this TODO resolved?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -26,29 +26,52 @@
 	public static final int DRIVETRAIN_ENCODER_FORWARD_RIGHT = 1;
+
+	
+	public static final boolean GEAR_SOLENOID_OUT_CONSTANT = true;
```
Remove _CONSTANT from these

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -26,29 +26,52 @@
 	
 	/**
+	 * Timing Constants
+	 * TODO: Set actual time
```
I'm guessing we find this via testing?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -26,29 +26,52 @@
+
+	public static final int CLIMBER_FORWARD_CONSTANT = 1;
+
+
```
::dies of whitespace poisoning::

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
Empty javadoc?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class GearIn extends Command {
+	
+	private GearCollector gearCollector = Robot.gearCollector;
```
Fix this inconsistent indentation; let's make sure we have a project standard (spaces? tabs?) & use that

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+    public GearIn() {
+    	requires(gearCollector);
+    	setInterruptible(true);
+        // Use requires() here to declare subsystem dependencies
```
I think these comments come by default with the class; can we remove them?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Sets the amount of the time the command runs for
+     */
+    // Called just before this Command runs the first time
```
Move this comment into the javadoc

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Solenoid is set to the "in" value and the command continues running until the timer runs out
+     */
+    // Called repeatedly when this Command is scheduled to run
```
Move this comment into the javadoc

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    // Make this return true when this Command no longer needs to run execute()
+    protected boolean isFinished() {
+        return isTimedOut();
```
This is definitely one finished condition; is there any other one? I see that Solenoid objects have a `get()` method that returns their current state, in or out, as a boolean. Do we need to run this command until timeout, or can we use the subsystem's `getGearSolenoid()` method to check to see if the command succeeded?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearIn.java:
```diff
@@ -0,0 +1,56 @@
+     * When the time passed exceeds the time established in initialize(), isTimedOut() returns true, 
+     *  making isFinished() return true and stopping the command
+     */
+    // Make this return true when this Command no longer needs to run execute()
```
Move this comment into the javadoc

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearOut.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class GearOut extends Command {
+	
+	private GearCollector gearCollector = Robot.gearCollector;
```
Fix inconsistent indentation

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearOut.java:
```diff
@@ -0,0 +1,56 @@
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
Eep, another empty Javadoc

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearOut.java:
```diff
@@ -0,0 +1,56 @@
+    public GearOut() {
+    	requires(gearCollector);
+    	setInterruptible(true);
+        // Use requires() here to declare subsystem dependencies
```
As with `GearIn`, we can probably ditch these comments

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/GearOut.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    // Make this return true when this Command no longer needs to run execute()
+    protected boolean isFinished() {
+        return isTimedOut();
```
Take a look at the comment for `GearIn`

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/GearCollector.java:
```diff
@@ -0,0 +1,51 @@
+
+	
+	private Solenoid gearSolenoid;
+	
```
https://drive.google.com/file/d/0Bwj8G2gowPqhVjJwTzV4NWRDc2c/view?usp=sharing

jameshochadel reviewed (changes requested):
Looks pretty good, but I'd like to see these things addressed before giving it the okay. Much of this is comments & whitespace, but there are a few items like the isFinished methods that'll require some consideration.

Commit:
Code review changes implemented

Commit:
Instatiate subsystems before OI

-Caused button to be unable to recognize subsystems their commands worked off of

Commit:
Gear collection is functional

-Modified code for a double-ended solenoid valve
-Assigned PCM to the correct address and solenoids to the correct ports
-Created command (used as default) that doesn't actuate either solenoid
-Created

Commit:
Corrected DriveTrain forward motion for left side

-fixed constants (left forward)
-updated firmware of talons and PCM
-multiply by forward constant within methods instead of using reverseOutput because that method only works on a closed loop
